### PR TITLE
key-export-abstract: Import/export with abstract encryption library

### PR DESCRIPTION
**Pull request by naure**
Issue #537.

This is a variant with an abstraction over encryption libraries.

### CODE REVIEW

File: src/keyshare_export.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+use crate::{errors::Result, keygen::KeySharePrivate};
+use std::{marker::PhantomData, ops::Deref};
+
+mod pke;
+pub use pke::Pke;
+
+mod sodium;
+pub use sodium::SodiumPke;
+
+/// Struct for handling the export and import of KeySharePrivate
```
Comment (gatoWololo) [inline src/keyshare_export.rs:12]: I think we should simplify this type a little bit. I don't think the generic `T` here is really buying us much here. Can we remove this generic?

File: src/keyshare_export.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+#[derive(Clone, Debug)]
+pub struct KeyShareEncrypted<T: Pke>(Vec<u8>, PhantomData<T>);
+
+impl<T: Pke> From<Vec<u8>> for KeyShareEncrypted<T> {
+    fn from(data: Vec<u8>) -> Self {
+        Self(data, PhantomData)
+    }
+}
+
+impl<T: Pke> From<KeyShareEncrypted<T>> for Vec<u8> {
```
Comment (gatoWololo) [inline src/keyshare_export.rs:22]: I think we should remove this `From` instance and simply provide a getter to the underlying data.

File: src/keyshare_export.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+pub use pke::Pke;
+
+mod sodium;
+pub use sodium::SodiumPke;
+
+/// Struct for handling the export and import of KeySharePrivate
+#[derive(Clone, Debug)]
+pub struct KeyShareEncrypted<T: Pke>(Vec<u8>, PhantomData<T>);
+
+impl<T: Pke> From<Vec<u8>> for KeyShareEncrypted<T> {
```
Comment (gatoWololo) [inline src/keyshare_export.rs:16]: I think we should remove this `From` instance and simply provide a constructor as an associated function: `fn from_vec(encrypted: Vec<u8>) -> Self`

File: src/keyshare_export.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    }
+}
+
+impl<T: Pke> From<KeyShareEncrypted<T>> for Vec<u8> {
+    fn from(encrypted: KeyShareEncrypted<T>) -> Vec<u8> {
+        encrypted.0
+    }
+}
+
+impl<T: Pke> Deref for KeyShareEncrypted<T> {
```
Comment (gatoWololo) [inline src/keyshare_export.rs:28]: I'm not sure if its necessary to have a `Deref` instance for this type. Could we remove it?

File: src/keyshare_export.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    ///     .expect("Failed to export keyshare");
+    /// let exported_bytes: Vec<u8> = exported.into();
+    ///
+    /// let to_import = KeyShareEncrypted::<SodiumPke>::from(exported_bytes);
+    /// let imported = to_import.import_keyshare(&sender_pk, &receiver_sk)
+    ///     .expect("Failed to import keyshare");
+    ///
+    /// assert_eq!(imported, key_share);
+    /// ```
+    pub fn export_keyshare(
```
Comment (gatoWololo) [inline src/keyshare_export.rs:72]: Per my other comment, on removing the generic on the struct. We will then re-introduce the generic here, at the function level: `fn export_keyshare<T: Pke>(...)`.

File: src/keyshare_export.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    ///
+    /// * `encrypted_key_share` - The encrypted byte vector of the
+    ///   KeySharePrivate.
+    /// * `sender_pk` - The public key of the sender for decryption.
+    /// * `receiver_sk` - The private key of the receiver for decryption.
+    ///
+    /// # Returns
+    ///
+    /// A `KeySharePrivate` object.
+    pub fn import_keyshare(
```
Comment (gatoWololo) [inline src/keyshare_export.rs:94]: Similar comment here, as my previous, about the generic.

**Review by gatoWololo — CHANGES_REQUESTED** [review]
Hello, we think we like this trait-based interface better. I left a few comments to simplify the code. But let me know if you disagree. Thanks!

File: src/keyshare_export.rs (hunk 1 above)
Comment (naure) [reply src/keyshare_export.rs:11]: Done.

File: src/keyshare_export.rs (hunk 2 above)
Comment (naure) [reply src/keyshare_export.rs]: Done.

File: src/keyshare_export.rs (hunk 3 above)
Comment (naure) [reply src/keyshare_export.rs]: Done.

**Review by becgabri — APPROVED** [review]
This looks good to me, thanks.

**Review by gatoWololo — APPROVED** [review]

**Pull request merged**